Add queue_empty and worker_exit hooks on the worker

jeremywadsack opened the pull request:
As discussed in #1647, to allow plugins and such to act on events such as the worker becoming idle (queue is now empty) and the worker exiting, this PR adds support for two new worker hooks.

  - `queue_empty` when the job queue empties and the worker becomes idle
  - `worker_exit` when the worker exits

Please review and let me know comments, thoughts. I'm happy to rename those if they don't feel "hook-y" enough. I wasn't sure, but also considered "after_queue_empty" and "before_exit".

jeremywadsack commented:
This test failure is related to an update to Airbrake. I can reproduce it on `master` with the latest Airbrake. I've narrowed the issue down to a change in `airbrake` gem between 8.1.4 and 8.2.0 which changed the dependency of `airbrake-ruby` from 3.2.6 to 4.1.0. My guess is it has something to do with airbrake/airbrake-ruby#429 or airbrake/airbrake-ruby#443. I tried configuring Airbrake to ignore the environment for the rake tests, but even though it accepts the configuration it still reports as being "closed".

jeremy commented on lib/resque/worker.rb:
```diff
@@ -11,6 +12,7 @@ module Resque
   # It also ensures workers are always listening to signals from you,
   # their master, and can react accordingly.
   class Worker
+    include Observable
```
How is Observable being used here? I'm not seeing changed/notify_observers.

jeremywadsack replied:
Good catch. I'll get that fixed up early next week,

jeremywadsack commented:
@jeremy Removed that spurious `include` and rebased so the tests pass. Let me know if you want me to squash commits.

jeremy commented on lib/resque/worker.rb:
```diff
@@ -1,3 +1,4 @@
+require 'observer'
```
Drop this require too?

jeremy commented:
Thanks! Looking good. One last observer require to nix. Please do squash 👍🏼

jeremywadsack commented:
Thanks @jeremy. Fixed that last issue and squashed. I also noticed that the test for "will call the queue empty hook only once" was not testing that it was called _just_ once, so I fixed that as well.

Let me know if you see anything else. Excited to have these hooks available!

alexcameron89 commented:
Awesome work here @jeremywadsack, I'm excited to take advantage of these new hooks. 👏 👏
